### PR TITLE
chore: replace AlecAivazis/survey with charmbracelet/bubbletea

### LINKED ISSUE #12720: AlecAivazis/survey is no longer maintained
https://github.com/AlecAivazis/survey currently shows this message:

image

We should consider migrating to another library. We only use it in `cli/cliui/select.go` at present, so it shouldn't be a very heavy lift.

### PR DESCRIPTION
**Pull request by DanielleMaywood**
fixes https://github.com/coder/coder/issues/12720

This PR replaces the unmaintained https://github.com/AlecAivazis/survey library with https://github.com/charmbracelet/bubbletea.

The existing tests for the components `Select`, `RichSelect` and `MultiSelect` do not test anything meaningful as the components short-circuit to return the first option when ran in a test and the tests just confirm this short-circuiting behaviour. Future work will be to remove this short-circuiting behaviour and fix the tests.

I've manually tested the behaviour of the components with:
- `go run ./cmd/cliui select`
  - `Select`
  - Before: https://asciinema.org/a/hPVc4qk95PK1qzCbMOWSDlP5a
  - After: https://asciinema.org/a/0Gzzhza6e3sleIhLnLvTBJSY5
- `go run ./cmd/coder exp prompt-example select`
  - `Select`
  - Before: https://asciinema.org/a/kKw91eDUrMF2uGh1hdusK3FQm
  - After: https://asciinema.org/a/RkNwryAJ2zf4W7z4m0kABu45I
- `go run ./cmd/coder exp prompt-example multiple`
  - `Select`, `MultiSelect`
  - Before: https://asciinema.org/a/kcQu4dfKLAMEcaMbD4nVdK5MN
  - After: https://asciinema.org/a/gR4SRwqWbuoyvLt1BpCOeYAKr
- `go run ./cmd/coder exp prompt-example multi-select`
  - `MultiSelect`
  - Before: https://asciinema.org/a/uIUyOw2ITEThRzk0JVwJfl7Kv
  - After: https://asciinema.org/a/LWo5r7VQ5X95wHlhWaowbyqXn
- `go run ./cmd/coder exp prompt-example rich-parameter`
  - `RichSelect`
  - Before: https://asciinema.org/a/yLPOdVKglxiEDdCzMn847O3Zl
  - After: https://asciinema.org/a/QCdmECsboRPWykhp2uKu4Yjgz
- `cdr templates delete` (against a local instance)
  - `Select`
  - Before: https://asciinema.org/a/2C3E4H5YtgDkgdRVY49aVGYHz
  - After: https://asciinema.org/a/W3kzC8R5r7T9o3NDH26lxl4FL
- `cdr orgs roles edit organization-auditor` (against a local instance)
  - `Select`, `MultiSelect`
  - Before: https://asciinema.org/a/cP3TTNmAjYGSy4RLt97W9bATo
  - After: https://asciinema.org/a/eKQlkL9C57Fl9VT7R9Y1Yx6Gu
- `cdr create -t devcontainer-docker` (against a local instance)
  - `RichSelect`
  - Before: https://asciinema.org/a/jLBTCWTIM5R5RTxX4OPwlsI3J
  - After: https://asciinema.org/a/aL6pJhnKEoIOorlCzd0Y4wyNs

### CODE REVIEW

File: cli/cliui/select.go
Comment (johnstcn) [inline cli/cliui/select.go:87]: nit: Is there any significance to this? Would it make sense to extract to a `var`?

File: cli/cliui/select.go
Comment (johnstcn) [inline cli/cliui/select.go:207]: nit: you could invert the condition on line 182 and return early e.g.

```
if m.selected != "" {
  selected := ...

  return s
}

// do the search
return s

```

File: cli/cliui/select.go
Comment (johnstcn) [inline cli/cliui/select.go:261]: Could we break the inner loop here?

File: cli/cliui/select.go
Comment (johnstcn) [inline cli/cliui/select.go:287]: What if `!ok`?

File: cli/cliui/select.go
Comment (johnstcn) [inline cli/cliui/select.go:319]: nit: invert and early return

File: cli/cliui/select.go
Comment (johnstcn) [inline cli/cliui/select.go:357]: Shouldn't this select all the options?

File: cli/cliui/select.go
Comment (johnstcn) [inline cli/cliui/select.go:384]: nit: use `strings.Builder` instead

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go:87]: The 7 comes from the previous default in the survey library https://github.com/AlecAivazis/survey/blob/160123ed7175f77bc8625b59c44b8dc03f5333e3/survey.go#L24

Maybe have a `const defaultModelHeight = 7` to make it less magic?

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go:357]: Oops my bad, not sure how I missed that

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go:287]: `ok` _should_ never be false here. Happy to return an error if that were to happen though.

File: cli/cliui/select.go
Comment (mafredri) [inline cli/cliui/select.go:100]: I don't think this can actually happen, but arguably, if !ok we probably shouldn't `return value, err` at the end. Perhaps we should handle !ok explicitly and return a new error?

File: cli/cliui/select.go
Comment (mafredri) [inline cli/cliui/select.go:104]: Nit: This could be simplified to `return m.selected, nil` instead of going through an intermediary variable.

File: cli/cliui/select.go
Comment (mafredri) [inline cli/cliui/select.go:121]: I found the blinking fairly distracting in the aciinemas, that behavior seemed different from the original. Is this what controls it and should we actually blink?

File: cli/cliui/select.go
Comment (mafredri) [inline cli/cliui/select.go:96]: Should we perhaps be using `tea.WithContext` here (`inv.Context()`)?

File: cli/cliui/select.go
Comment (mafredri) [inline cli/cliui/select.go:128]: What happens if the program receives a `kill -INT $PID` while tea is active? Do we handle it nicely?

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go:326]: I've slightly changed the logic here. I just noticed it is slightly different to how the `func (m selectModel) Update` function works. `m.search.Update` can handle more than just a `tea.KeyMsg`. With the updated logic does this still make sense?

File: cli/cliui/select.go
Comment (mafredri) [inline cli/cliui/select.go]: Suggestion: Invert for early return and reduced indentation.

**Review by mafredri — COMMENTED** [review]
Nice work, and thanks for the asciinemas! Left a few comments on changes I'd like to see in all functions/models, but otherwise looks good.

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go:121]: Yes, this is what controls it https://pkg.go.dev/github.com/charmbracelet/bubbles@v0.19.0/textinput#Blink.

I'm happy to back it out

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go:96]: Oh that's a good spot, yes I think we should

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go]: Have changed this to error now if `!ok`

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go:98]: Added some error handling

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go]: It shouldn't but I have added some error handling now

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go]: Good spot, yes I've done this now

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go]: Have done!

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go:135]: Unless I'm doing something wrong, it appears nothing happens?

Comment (mafredri) [reply cli/cliui/select.go:135]: Hmm, did you try with a `coder` command or cliui test binary? With `coder` I would expect the program to exit with a cancelled.

Comment (DanielleMaywood) [reply cli/cliui/select.go:135]: I used the `cliui` binary with `go run ./cmd/cliui select`. I shall try against the coder binary

File: cli/cliui/select.go
Comment (mafredri) [inline cli/cliui/select.go:99]:
```suggestion
```

Nit: Avoid separating error handling with assignment. It's makes it easier to stick logic in-between.

File: cli/cliui/select.go
Comment (mafredri) [inline cli/cliui/select.go:113]:
```suggestion
	return model.selected, nil
```
Nit: Only return err if it wasn't handled.

**Review by johnstcn — APPROVED** [review]
Nice work and thorough testing!

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go:113]: My bad, thanks for spotting that

File: cli/cliui/select.go
Comment (mafredri) [reply cli/cliui/select.go]: Cool, thanks. I'm guessing the `Update` method is actually capable of swapping out the model, so as long as one takes care not to do it, yeah, shouldn't happen. But better safe than sorry. 😄

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go:136]: Ah so running it without `go run` changes the behaviour. If the caller doesn't handle the `Canceled` error then it returns an empty string for the `Select` component

Comment (mafredri) [reply cli/cliui/select.go:136]: > Ah so running it without go run changes the behaviour.
Yeah go run compiles the binary and it runs with a different PID than the actual `go run` command.
> If the caller doesn't handle the Canceled error then it returns an empty string for the Select component
This is what I would expect 👍🏻

Comment (DanielleMaywood) [reply cli/cliui/select.go:136]: Have just checked on the `MultiSelect` component, it returns the currently selected options.

**Review by mafredri — APPROVED** [review]
LGTM after the conflicts are resolved, nice work!

File: cli/cliui/select.go
Comment (mafredri) [reply cli/cliui/select.go:136]: Does it populate error as well (it should)?

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go:129]: Doing some more research, it appears that the `textinput` component also triggers the blinking behaviour

File: cli/cliui/select.go
Comment (DanielleMaywood) [reply cli/cliui/select.go:136]: Just tested and it does not! Need to handle that then, `err` is `nil`

Comment (DanielleMaywood) [reply cli/cliui/select.go:136]: I can confirm by disabling the builtin signal handler it is possible to get `err` to contain the `Canceled` error

Comment (mafredri) [reply cli/cliui/select.go:136]: Might I suggest returning a func instead of the channel for close/deregistering the handler? I feel the usage will be a bit more idiomatic/easy to understand.

On the other hand, perhaps we don't need to handle signals at all and leave it to be the responsibility of the `coder` command that needs it? I guess it's possible the terminal is left in a bad state unless the signal is handled though (not sure if it enters raw mode or not).

PS. You might want to only handle `signal.Interrupt` for the Windows compatibility.

Comment (DanielleMaywood) [reply cli/cliui/select.go:136]: Yes, definitely a good idea. Have made that change 👍

Killing the process with `kill -TERM $PID` leaves the cursor in a bad state if we don't explicitly handle it (`syscall.SIGTERM`).

Will push these changes up, can always revert if unhappy.

**Review by mtojek — APPROVED** [review]
LGTM 👍 Feel free to merge once you update `flake.nix` `go.mod` and make CI happy. Nice contribution!

**Comment by Emyrk** [comment]
Late to the party, but one thing that bothered me with the `survey` package was that the prompt question remained after it was answered. I wonder if we can fix that now.

**Comment by DanielleMaywood** [comment]
> Late to the party, but one thing that bothered me with the `survey` package was that the prompt question remained after it was answered. I wonder if we can fix that now.

This absolutely can be changed now! Personally I prefer the question (with the answer) remaining, so I'd be curious if there is any way of finding out if that would be a positive/negative change amongst our customers.

**Pull request merged**
